src/middlewared/middlewared/plugins/smb_/util_passdb.py

```diff
@@ -0,0 +1,543 @@
+
+    This provides a convenient query API for passdb entries. Wraps around
+    passdb_entries() and same failure scenarios apply.
+
+    Params:
+        filters - standard query-filters
+        options - standard query-options
+
+    Returns:
+        filterable returns with asdict() output of PDBEntry objects
+
+    Raises:
+       PassdbMustReinit - internal inconsistencies in passdb file
+    """
+    try:
+        return filter_list(passdb_entries(as_dict=True), filters, options)
+    except FileNotFoundError:
+        return []
+
+
+def insert_passdb_entries(entries: list[PDBEntry]) -> None:
+    """ Insert multiple groupmap entries under a transaction lock
+
+    Each PDBEntry requires two TDB insertions and so the entire list
+    is submitted under a TDB transaction lock, which means that in case
+    of failure changes are rolled back to the state prior to insertion.
+
+    Entries that already exist will be overwritten.
+
+    Params:
+        entries - list of PDBEntry objects to be inserted
+
+    Raises:
+        TypeError - list item isn't PDBEntry object
+        RuntimeError - TDB library error
+    """
+
+    if not os.path.exists(PASSDB_PATH):
+        _add_version_info()
+
+    batch_ops = []
+
+    for entry in entries:
+        if not isinstance(entry, PDBEntry):
+            raise TypeError(f'{type(entry)}: not a PDBEntry')
+
+        samu_data = _pack_pdb_entry(entry)
+        batch_ops.extend([
+            TDBBatchOperation(
+                action=TDBBatchAction.SET,
+                key=f'{USER_PREFIX}{entry.username}',
+                value=b64encode(samu_data)
+            ),
+            TDBBatchOperation(
+                action=TDBBatchAction.SET,
+                key=f'{RID_PREFIX}{entry.user_rid:08x}',
+                value=b64encode(entry.username.encode() + b'\x00')
+            )
+        ])
+
+    if len(batch_ops) == 0:
+        # nothing to do, avoid taking lock
+        return
+
+    with get_tdb_handle(PASSDB_PATH, PASSDB_TDB_OPTIONS) as hdl:
+        hdl.batch_op(batch_ops)
+
+
+def delete_passdb_entry(username: str, rid: int) -> None:
+    """ Delete a passdb entry under a transaction lock """
+    if not os.path.exists(PASSDB_PATH):
+        # passdb.tdb doesn't exist so nothing to do
+        return
+
+    with get_tdb_handle(PASSDB_PATH, PASSDB_TDB_OPTIONS) as hdl:
+        # Do this under transaction lock to force atomicity of changes
+        try:
+            hdl.batch_op([
+                TDBBatchOperation(
+                    action=TDBBatchAction.DEL,
+                    key=f'{USER_PREFIX}{username}'
+                ),
+                TDBBatchOperation(
+                    action=TDBBatchAction.DEL,
+                    key=f'{RID_PREFIX}{rid:08x}'
+                )
+            ])
+        except RuntimeError:
+            # entries do not exist
+            pass
+
+
+def update_passdb_entry(entry: PDBEntry) -> None:
+    """ Update an existing passdb entry
```

Comment:
Hmmm.  Again are we _only_ doing an update (and not an insert)?  That seems to be at odds with handling a missing `PASSDB_PATH`.  Seems like its insert or update.